src/Expander/Monad.hs

```diff
@@ -688,3 +708,106 @@ datatypeInfo datatype = do
     Nothing ->
       throwError $ InternalError $ "Unknown datatype " ++ show datatype
     Just info -> pure info
+
+bind :: Binding -> EValue -> Expand ()
+bind b v =
+  modifyState $
+  over expanderExpansionEnv $
+  \(ExpansionEnv env) ->
+    ExpansionEnv $ Map.insert b v env
+
+-- | Add a binding to the current module's table
+addBinding :: Ident -> Binding -> BindingInfo SrcLoc -> Expand ()
+addBinding (Stx scs _ name) b info = do
+  modifyState $ over (expanderCurrentBindingTable . at name) $
+    (Just . ((scs, b, info) :) . fromMaybe [])
+
+addImportBinding :: Ident -> Binding -> Expand ()
+addImportBinding x@(Stx _ loc _) b =
+  addBinding x b (Imported loc)
+
+addDefinedBinding :: Ident -> Binding -> Expand ()
+addDefinedBinding x@(Stx _ loc _) b =
+  addBinding x b (Defined loc)
+
+addLocalBinding :: Ident -> Binding -> Expand ()
+addLocalBinding x@(Stx _ loc _) b =
+  addBinding x b (BoundLocally loc)
+
+addModuleScope :: HasScopes a => a -> Expand a
+addModuleScope stx = do
+  mn <- view (expanderLocal . expanderModuleName) <$> ask
+  sc <- moduleScope mn
+  return $ addScope' stx sc
+
+-- | Add the current phase's root scope at the current phase
+addRootScope :: HasScopes a => a -> Expand a
+addRootScope stx = do
+  p <- currentPhase
+  rsc <- phaseRoot
+  return (addScope p stx rsc)
+
+-- | Add the current phase's root scope at all phases (for binding occurrences)
+addRootScope' :: HasScopes a => a -> Expand a
+addRootScope' stx = do
+  rsc <- phaseRoot
+  return (addScope' stx rsc)
+
+-- | Schedule an expression expansion task, returning the pointer to
+-- which it will be written.
+schedule :: Ty -> Syntax -> Expand SplitCorePtr
+schedule t stx@(Syntax (Stx _ loc _)) = do
+  dest <- liftIO newSplitCorePtr
+  saveOrigin dest loc
+  forkExpandSyntax (ExprDest t dest) stx
+  return dest
+
+kernelExports :: Expand Exports
+kernelExports = view expanderKernelExports <$> getState
+
+completely :: Expand a -> Expand a
+completely body = do
+  oldTasks <- getTasks
+  clearTasks
+  a <- body
+  remainingTasks <- getTasks
+  unless (null remainingTasks) $ do
+    throwError (NoProgress (map (view _3) remainingTasks))
+  setTasks oldTasks
+  pure a
+
+getTasks :: Expand [(TaskID, ExpanderLocal, ExpanderTask)]
+getTasks = view expanderTasks <$> getState
+
+setTasks :: [(TaskID, ExpanderLocal, ExpanderTask)] -> Expand ()
+setTasks = modifyState . set expanderTasks
+
+clearTasks :: Expand ()
+clearTasks = modifyState $ set expanderTasks []
```

Comment:
Did you mean to move `getTasks` and `clearTasks` here rather than copy them? `getTasks` and `clearTasks` are still in `src/Expander.hs`.